### ISSUE TITLE
Masked mean for advantage normalization?

### ISSUE THREAD

**Issue by finnBsch** [MarcoMeter/recurrent-ppo-truncated-bptt#10]
Hi, 
Why is the masked mean not used for advantage normalization? To my understand, it is padded too, no?

**Comment by MarcoMeter** [comment]
Hey, good catch! We will check whether we will normalize all advantages before padding or whether we apply masked_mean and masked_std on mini batch level.

**Comment by finnBsch** [comment]
Some papers suggest batch-wise standardization to be superior but obviously takes some more effort to do a masked_mean and masked_std every update. Does pytorch support any efficient implementations for this?

**Comment by MarcoMeter** [comment]
The advantage is now correctly normalized. Based on our experiments, we did not observe any drops or gains in performance. The same applies to normalizing the advantages across the entire batch.

**Comment by finnBsch** [comment]
Generally, PyTorch allows to select by binary mask. I haven't ran extensive tests but it seems to improve performance over the `masked_mean` function provided in the trainer. I can send a pull request for that if you want to adopt the changes I have made.

**Comment by MarcoMeter** [comment]
Are you referring to `masked_select()` ? Feel free to make a PR. We can profile alternatives.

Here are some results concerning the advantage normalization. Apparently, the performance is best if advantages are not normalized.

![output](https://user-images.githubusercontent.com/6996955/205065641-259e6a51-2e45-4f39-9a00-cbf5698b6822.png)

We ran experiments on the Mystery Path Grid environment (goal and origin visible). Each curve shows the interquartile mean success rate and a bootstrapped confidence interval of 95% on 5 training runs. [Details of the environment can be found here.](https://openreview.net/forum?id=jHc8dCx6DDr)

**Comment by finnBsch** [comment]
I will do it later today.

Your results are interesting, I have experienced the same. I am using my recurrent PPO implementation (which is inspired mainly by yours) to solve a game/scenario which differs a lot from the one you used. I can run similar tests there to check performance. While the general results are very good, I am having trouble however to get it to train as smoothly as the results you are showing. Furthermore, training seems to take forever. With 32 agents, 400 steps per episode, I need 20.000 episodes for good results (~256 million steps). Takes about 10 hours on my RTX 3070, so it might take a while.

**Comment by MarcoMeter** [comment]
It looks like masked_select is faster.

Code for profiling both functions
```python
import cProfile, pstats
import torch

def masked_mean(tensor:torch.Tensor, mask:torch.Tensor) -> torch.Tensor:
    return (tensor.T * mask).sum() / torch.clamp((torch.ones_like(tensor.T) * mask).float().sum(), min=1.0)

def select_masked_mean(tensor, mask):
    return torch.masked_select(tensor, mask).mean()

def main():
    profiler = cProfile.Profile()
    profiler.enable()
    for i in range(100000):
        masked_mean(torch.zeros((1024), dtype=torch.float32), torch.ones((1024), dtype=torch.bool))
    profiler.disable()
    stats = pstats.Stats(profiler).sort_stats('cumtime')
    stats.print_stats()

if __name__ == "__main__":
    main()
```

Results
```
ncalls  tottime  percall  cumtime  percall filename:lineno(function)
   100000    2.294    0.000    3.771    0.000 run_profile.py:4(masked_mean)
```
```
ncalls  tottime  percall  cumtime  percall filename:lineno(function)
   100000    0.098    0.000    2.230    0.000 run_profile.py:8(select_masked_mean)
```

**Comment by MarcoMeter** [comment]
I'm closing this issue now. Given the latest update, advantages are not padded anymore. We did not change the advantage normalization. We are not claiming that this is the best way for every environment, but this can be easily modified to the users' liking.
